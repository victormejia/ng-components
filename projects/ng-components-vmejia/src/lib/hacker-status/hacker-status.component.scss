$red: #c0392b;
$green: #27ae60;
$yellow: #f1c40f;

.status-pulse {
  position: relative;
  width: 10px;
  height: 10px;

  &:hover {
    cursor: pointer;
  }
}

.pulse {
  width: 10px;
  height: 10px;
  border-radius: 30px;
  position: absolute;

  &.red {
    background-color: $red;
  }

  &.green {
    background-color: $green;
  }

  &.yellow {
    background-color: $yellow;
  }
}

.dot {
  background: transparent;
  border-radius: 60px;
  height: 20px;
  width: 20px;
  position: absolute;
  top: -10px;
  left: -10px;
  opacity: 0;
  border-width: 5px;
  border-style: solid;

  &.red {
    border-color: $red;
    animation: pulse 2s ease-out;
    animation-iteration-count: infinite;
  }

  &.green {
    border: 10px solid $green;
  }

  &.yellow {
    border: 10px solid $yellow;
  }
}

@keyframes pulse {
 0% {
    transform: scale(0);
    opacity: 0.0;
 }
 10% {
    transform: scale(0.1);
    opacity: 0.1;
 }
 20% {
    transform: scale(0.2);
    opacity: 0.2;
 }
 30% {
    transform: scale(0.3);
    opacity: 0.3;
 }
 40% {
    transform: scale(0.4);
    opacity: 0.4;
 }
 50% {
    transform: scale(0.5);
    opacity: 0.5;
 }
 60% {
    transform: scale(0.6);
    opacity: 0.6;
 }
 70% {
    transform: scale(0.75);
    opacity: 0.5;
 }
 80% {
    transform: scale(0.8);
    opacity: 0.4;
 }
 90% {
    transform: scale(0.85);
    opacity: 0.2;
 }
 100% {
    transform: scale(0.9);
    opacity: 0.0;
 }
}
