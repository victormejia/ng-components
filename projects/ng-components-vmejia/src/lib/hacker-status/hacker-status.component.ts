import { Component, OnInit, Input } from '@angular/core';

@Component({
  selector: 'vm-hacker-status',
  template: `
    <div class="status-pulse">
      <span class="pulse" [ngClass]="color"></span>
      <span class="dot" [ngClass]="color"></span>
    </div>
  `,
  styleUrls: ['./hacker-status.component.scss']
})
export class HackerStatusComponent implements OnInit {

  @Input() status: string;
  color: string;

  constructor() { }

  ngOnInit() {
    this.color = this.statusToColor(this.status);
  }

  statusToColor(status: string) {
    switch (status) {
      case 'danger':
        return 'red';
      case 'safe':
        return 'green';
      case 'warning':
        return 'yellow';
      default:
        return 'yellow';
    }
  }

}
