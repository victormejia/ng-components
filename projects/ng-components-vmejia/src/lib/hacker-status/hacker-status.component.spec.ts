import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { HackerStatusComponent } from './hacker-status.component';
import { Component } from '@angular/core';
import { By } from '@angular/platform-browser';

@Component({
  template: `
    <vm-hacker-status [status]="appStatus"></vm-hacker-status>
  `
})
class TestHostComponent {
  appStatus: string;
}

describe('HackerStatusComponent', () => {
  let testHost: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ TestHostComponent, HackerStatusComponent ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    testHost = fixture.componentInstance;
  });

  it('should create', () => {
    expect(testHost).toBeTruthy();
  });

  it('should set pulse color to green when input is "safe"', () => {
    testHost.appStatus = 'safe';

    fixture.detectChanges();

    const pulseEl: HTMLElement = fixture.debugElement.query(By.css('.pulse')).nativeElement;

    expect(pulseEl.classList).toContain('green');
  });

  it('should set pulse color to green when input is "safe" (snapshot testing)', () => {
    testHost.appStatus = 'safe';

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should set pulse color to red when input is "danger" (snapshot testing)', () => {
    testHost.appStatus = 'danger';

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should set pulse color to yellow when input is "warning" (snapshot testing)', () => {
    testHost.appStatus = 'warning';

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should set pulse color to yellow when input is undefined/null (snapshot testing)', () => {
    testHost.appStatus = null;

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
