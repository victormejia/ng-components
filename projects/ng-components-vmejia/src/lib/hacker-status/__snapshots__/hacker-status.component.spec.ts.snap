// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackerStatusComponent should set pulse color to green when input is "safe" (snapshot testing) 1`] = `
<ng-component
  appStatus={[Function String]}
>
  <vm-hacker-status
    ng-reflect-status="safe"
  >
    <div
      class="status-pulse"
    >
      <span
        class="pulse green"
        ng-reflect-klass="pulse"
        ng-reflect-ng-class="green"
      />
      <span
        class="dot green"
        ng-reflect-klass="dot"
        ng-reflect-ng-class="green"
      />
    </div>
  </vm-hacker-status>
</ng-component>
`;

exports[`HackerStatusComponent should set pulse color to red when input is "danger" (snapshot testing) 1`] = `
<ng-component
  appStatus={[Function String]}
>
  <vm-hacker-status
    ng-reflect-status="danger"
  >
    <div
      class="status-pulse"
    >
      <span
        class="pulse red"
        ng-reflect-klass="pulse"
        ng-reflect-ng-class="red"
      />
      <span
        class="dot red"
        ng-reflect-klass="dot"
        ng-reflect-ng-class="red"
      />
    </div>
  </vm-hacker-status>
</ng-component>
`;

exports[`HackerStatusComponent should set pulse color to yellow when input is "warning" (snapshot testing) 1`] = `
<ng-component
  appStatus={[Function String]}
>
  <vm-hacker-status
    ng-reflect-status="warning"
  >
    <div
      class="status-pulse"
    >
      <span
        class="pulse yellow"
        ng-reflect-klass="pulse"
        ng-reflect-ng-class="yellow"
      />
      <span
        class="dot yellow"
        ng-reflect-klass="dot"
        ng-reflect-ng-class="yellow"
      />
    </div>
  </vm-hacker-status>
</ng-component>
`;

exports[`HackerStatusComponent should set pulse color to yellow when input is undefined/null (snapshot testing) 1`] = `
<ng-component
  appStatus="null"
>
  <vm-hacker-status>
    <div
      class="status-pulse"
    >
      <span
        class="pulse yellow"
        ng-reflect-klass="pulse"
        ng-reflect-ng-class="yellow"
      />
      <span
        class="dot yellow"
        ng-reflect-klass="dot"
        ng-reflect-ng-class="yellow"
      />
    </div>
  </vm-hacker-status>
</ng-component>
`;
